### PR TITLE
serverutils: provide SQLConn/SQLConnE in ApplicationLayerInterface

### LINKED ISSUE #107672: serverutils: make a helper function to open a SQL conn with a differnt username
This would be an extension of #107659

We have plenty of tests that do not like the `*gosql.DB` returned by `StartServer` and instead do:

1. construct a URL object from the server's SQLAddr with a custom username
2. run gosql.Open

That's about 4-5 lines of repetitive code. Instead we should have a helper "open a connection for a given user".

Epic: CRDB-18499

Jira issue: CRDB-30153

### PR DESCRIPTION
**Pull request by knz**
Fixes  #107672.
Part of solving #107058.
Informs #106772.

Epic: CRDB-18499

### CODE REVIEW

**Comment by knz** [comment]
TODO: remove StorageClusterConn

File: pkg/server/testserver_sqlconn.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+			cleanupFn()
+			return nil, err
+		}
+	}
+
+	// Ensure the connection is closed at the end of the test.
+	stopper.AddCloser(stop.CloserFn(func() {
+		err := goDB.Close()
+		cleanupFn()
+		log.Infof(context.Background(), "closed test SQL connection: %v", err)
```
Comment (stevendanna) [inline pkg/server/testserver_sqlconn.go:93]: Perhaps only log this on error. Up to you.

**Review by stevendanna — APPROVED** [review]
Nice. 

I reviewed the testserver/testcluster changes, the backupccl changes, and then skimmed the other packages.

File: pkg/server/testserver_sqlconn.go (hunk 1 above)
Comment (knz) [reply pkg/server/testserver_sqlconn.go]: Done.

**Comment by knz** [comment]
bors r=stevendanna

**Comment by stevendanna** [comment]
We should consider backporting this or it will be hard to backport other test-related fixes.

**Review by yuzefovich — COMMENTED** [review]
Nice cleanup!

> Reviewed 4 of 30 files at r1, 2 of 50 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107659#-:-NaT6G9kAaDPnveqZHzI:b7oy6or)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Pull request merged**